vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -211,142 +242,194 @@ type
+
+
+    member self.OnAfterLoadProject (args:Events.LoadProjectEventArgs) = self.SetupProjectHierarchy args.RealHierarchy
+
+
+    /// Get the exact options for a single-file script
+    member __.ComputeSingleFileOptions (fileName, loadTime, fileContents, hierarchy, workspace: Workspace) = async {
+        let extraProjectInfo = Some (box workspace)
+        if SourceFile.MustBeSingleFileProject fileName then 
+        //if isScriptFile fileName then 
+            let! _options, _diagnostics = projectInfoManager.Checker.GetProjectOptionsFromScript (fileName, fileContents, loadTime, [| |], ?extraProjectInfo=extraProjectInfo) 
+            //ProjectOptionsCache.fsxProjectOptions fileName (Some workspace)
+            return projectInfoManager.GetProjectOptionsForProjectSite ( fileName, loadTime,hierarchy,extraProjectInfo)
+        else
+            //ProjectOptionsCache.singleFileProjectOptions fileName
+            return projectInfoManager.GetProjectOptionsForProjectSite ( fileName, loadTime,hierarchy, extraProjectInfo)
+    }
+
+
+    member self.ProjectTracker : VisualStudioProjectTracker =
+        match projectTracker with
+        | Some tracker -> tracker 
+        | None -> 
+            let tracker = self.Workspace.GetProjectTrackerAndInitializeIfNecessary Shell.ServiceProvider.GlobalProvider 
+            projectTracker <- Some tracker
+            tracker 
+
+
+    override self.Initialize () =
+        base.Initialize ()
+
+        let projectTracker = (self.ProjectTracker: VisualStudioProjectTracker) 
+        let solutionVS = package.GetService<SVsSolution,IVsSolution>()
+        self.Workspace.AdviseSolutionEvents solutionVS
+        
+        self.Workspace.Options <- self.Workspace.Options.WithChangedOption (Completion.CompletionOptions.BlockForCompletionItems, FSharpConstants.FSharpLanguageName, false)
+        self.Workspace.Options <- self.Workspace.Options.WithChangedOption (Shared.Options.ServiceFeatureOnOffOptions.ClosedFileDiagnostic, FSharpConstants.FSharpLanguageName, Nullable false)
+
+        self.Workspace.DocumentClosed.Add (fun args ->
+           tryRemoveSingleFileProject args.Document.Project.Id 
+        )
+        
+        self.Workspace.DocumentOpened.Add (fun args ->
+            projectInfoManager.AddProject <| getProject args.Document.Project.Id
+        )
+
+        self.Workspace.WorkspaceChanged.Add self.OnWorkspaceChanged
+        Events.SolutionEvents.OnAfterLoadProject.Add self.OnAfterLoadProject
+        
+        let setupProjects () = async {
+            do! Async.AwaitEvent Events.SolutionEvents.OnAfterOpenSolution |> Async.Ignore
+            use _ = Events.SolutionEvents.OnAfterOpenProject |> Observable.subscribe ( fun args ->
+                self.SetupProjectHierarchy args.Hierarchy |> ignore
+            )
+            for project in solutionDTE.GetProjects () do 
+                self.SetupProject project |> getProject |> projectInfoManager.AddProject
```

Comment:
If I remember correctly performance was worse with projects being setup during solution load. We don't want this the first time and also we want to avoid it on subsequent solution openings.